README
=== README for Geo-KML version 0.95
=   Generated on Fri Aug  8 17:20:09 2014 by OODoc 2.01

There are various ways to install this module:

 (1) if you have a command-line, you can do:
       perl -MCPAN -e 'install <any package from this distribution>'

 (2) if you use Windows, have a look at http://ppm.activestate.com/

 (3) if you have downloaded this module manually (as root/administrator)
       gzip -d Geo-KML-0.95.tar.gz
       tar -xf Geo-KML-0.95.tar
       cd Geo-KML-0.95
       perl Makefile.PL
       make          # optional
       make test     # optional
       make install

For usage, see the included manual-pages or
    http://search.cpan.org/dist/Geo-KML-0.95/

Please report problems to
    http://rt.cpan.org/Dist/Display.html?Queue=Geo-KML



ChangeLog

==== version history for Geo::KML

version 0.95: Fri Aug  8 17:19:44 CEST 2014

	Changes:

	- disable support for this module, when XML::Compile > 1.43
	  final release.

version 0.94: Tue Aug 28 09:09:50 CEST 2012

	Fixes:

	- another attempt to create a wrapper around Archive::Zip to
	  be able to handle kmz data from strings.

	Improvements:
	- add schema for kml2.0

version 0.93: Thu Jun  9 15:54:21 CEST 2011

	Improvements:

	- Geo::KML::from() now also accepts kmz files which are passed as
	  scalars and strings.

	- added some example files and scripts.

	- documentation updates.

version 0.92: Fri Dec 24 11:15:46 CET 2010

	Fixes:

	- link typo in documentation.
	  rt.cpan.org #63477 [Frank Wiegand]

	Improvements:

	- colors only did accept integers, like 0x123456, but now also
	  accept strings like '123456'

version 0.91: Fri Sep 24 21:46:28 CEST 2010

	Changes:

	- hook color code processing, to avoid working with binary
	  strings for colors.

version 0.90: Thu Sep 23 10:10:13 CEST 2010

	Changes:

	- writeKML() returns the mime-type of the produced data.

	- do not automatically add a "Document" root element. There
	  are more options in the top node.

	Improvements:

	- add MIME_KML and MIME_KMZ to Geo::KML::Util.

	- add gx extensions to KML 2.2.0

	- nicer handling of floats, integers and mixed content.

	- add test for reading 2.2.0

	- move t/99pod.t to xt, which removes the requirement for
	  Test::Pod

	- jump version number to show increased usability.

	- remove dependency to Test::Deep

	- better docs

version 0.03: Thu Mar 19 09:21:11 CET 2009

	Fixes:

	- do not use /bin/pwd in t/pod.t

	- use Archive::Tar::addString(), not ::addMember() do support older
	  versions of Archive::Tar

	Improvements

	- deprecated readKML()

	- implement from() as nicer alternative to readKML()

version 0.02: Mon Aug 11 15:37:15 CEST 2008

	Fixes:

	- ISA fixes in the writer.

	Improvements:

	- use XML::Compile::Cache as base

	- no need for Document top-element in the writer.

	- reuse created default reader objects.

version 0.01: Mon Jun 16 13:00:30 CEST 2008
	- initial release


Makefile.PL
use ExtUtils::MakeMaker;

use 5.008;

eval "require XML::Compile";
if(!$@)
{   my $v = $XML::Compile::VERSION || 1.43;

    # This limitation will not be lifted
    die "Only works with XML::Compile <= 1.43, not $v" if $v ge 1.44;
}

WriteMakefile
 ( NAME	 	=> 'Geo::KML'
 , VERSION	=> '0.95'
 , PREREQ_PM    =>
   { XML::Compile        => 1.18
   , XML::Compile::Cache => 0.97
   , Log::Report         => 0.28
   , Archive::Zip        => 1.30
   , Test::More          => 0.54
   , Data::Peek          => 0.32
   , File::Temp          => 0
   }
 , EXE_FILES    => [ ]
 , AUTHOR       => 'Mark Overmeer'
 , ABSTRACT     => 'Create GoogleEarth and -Map KML/KMZ'
 , LICENSE      => 'perl'
 );

### used by oodist during production of distribution
sub MY::postamble { <<'__POSTAMBLE' }

# for DIST
RAWDIR          = ../public_html/geo-kml/raw
DISTDIR         = ../public_html/geo-kml/source
LICENSE         = artistic

# for POD
FIRST_YEAR      = 2008
EMAIL           = perl@overmeer.net
WEBSITE         = http://perl.overmeer.net/geo-kml
PODTAIL		= ../Geo-shared-podtail
__POSTAMBLE


lib/Geo/KML/Util.pm
# Copyrights 2008-2014 by [Mark Overmeer].
#  For other contributors see ChangeLog.
# See the manual pages for details on the licensing terms.
# Pod stripped from pm file by OODoc 2.01.
use warnings;
use strict;

package Geo::KML::Util;
use vars '$VERSION';
$VERSION = '0.95';

use base 'Exporter';

use Log::Report 'geo-kml';

my @kml20   = qw/NS_KML_20/;
my @kml21   = qw/NS_KML_21/;
my @kml22b  = qw/NS_KML_22BETA NS_ATOM_2005 NS_XAL_20/;
my @kml220  = qw/NS_KML_22 NS_KML_220 NS_ATOM_2005 NS_XAL_20 NS_KML_EXT_22/;
my @mime    = qw/MIME_KML MIME_KMZ/;

our @EXPORT = (@kml20, @kml21, @kml22b, @kml220, @mime);

our %EXPORT_TAGS =
  ( kml20     => \@kml20
  , kml21     => \@kml21
  , kml22beta => \@kml22b
  , kml220    => \@kml220
  , mime      => \@mime
  );


use constant NS_KML_20     => 'http://earth.google.com/kml/2.0';


use constant NS_KML_21     => 'http://earth.google.com/kml/2.1';


use constant NS_KML_22BETA => 'http://earth.google.com/kml/2.2';


use constant
  { NS_KML_22     => 'http://www.opengis.net/kml/2.2'
  , NS_KML_220    => 'http://www.opengis.net/kml/2.2'
  , NS_ATOM_2005  => 'http://www.w3.org/2005/Atom'
  , NS_XAL_20     => 'urn:oasis:names:tc:ciq:xsdschema:xAL:2.0'
  , NS_KML_EXT_22 => 'http://www.google.com/kml/ext/2.2'
  };


use constant
 { MIME_KML => 'application/vnd.google-earth.kml+xml'
 , MIME_KMZ => 'application/vnd.google-earth.kmz'
 };

1;


lib/Geo/KML.pm
# Copyrights 2008-2014 by [Mark Overmeer].
#  For other contributors see ChangeLog.
# See the manual pages for details on the licensing terms.
# Pod stripped from pm file by OODoc 2.01.

use warnings;
use strict;

package Geo::KML;
use vars '$VERSION';
$VERSION = '0.95';

use base 'XML::Compile::Cache';

use Log::Report 'geo-kml', syntax => 'SHORT';

use Geo::KML::Util;    # all constants
use XML::Compile::Util qw/pack_type type_of_node/;
use XML::Compile       ();
use Archive::Zip       qw/AZ_OK COMPRESSION_LEVEL_DEFAULT/;
use Data::Peek         qw/DDual/;
use File::Temp         qw/tempfile/;
use Fcntl              qw/SEEK_SET/;

use Data::Dumper;

use constant KML_NAME_IN_KMZ => 'doc.kml';

my %ns2version  =
  ( &NS_KML_20     => '2.0'
  , &NS_KML_21     => '2.1'
  , &NS_KML_22BETA => '2.2-beta'
  , &NS_KML_22     => '2.2.0'
  );
my %version2ns  = reverse %ns2version;
my %implement;

my %info =
  ( '2.0'   =>
    { prefixes => [ '' => NS_KML_20 ]
    , schemas  => [ 'kml-2.0/*.xsd' ]
    }

  , '2.1'   =>
    { prefixes => [ '' => NS_KML_21 ]
    , schemas  => [ 'kml-2.1/*.xsd' ]
    }

  , '2.2-beta' =>
    { prefixes => [ '' => NS_KML_22BETA, atom => NS_ATOM_2005, xal=> NS_XAL_20]
    , schemas  => [ 'kml-2.2-beta/kml22beta.xsd', 'kml-2.2-beta/fixes.xsd'
                  , 'atom-2005/*.xsd', 'xal-2.0/*.xsd' ]
    }

  , '2.2.0' =>
    { prefixes => [ '' => NS_KML_220, atom => NS_ATOM_2005, xal => NS_XAL_20
                  , gx => NS_KML_EXT_22 ]
    , schemas  => [ 'kml-2.2.0/*.xsd', 'atom-2005/*.xsd', 'xal-2.0/*.xsd' ]

    , hooks_r  => [ { type => 'colorType', replace => \&color_hex_read } ]
    , hooks_w  => [ { type => 'colorType', replace => \&color_hex_write} ]
    }
  );


sub init($)
{   my ($self, $args) = @_;

    my $version  =  $args->{version}
        or error __x"KML object requires an explicit version";

    unless(exists $info{$version})
    {   exists $ns2version{$version}
            or error __x"KML version {v} not recognized", v => $version;
        $version = $ns2version{$version};
    }
    $self->{GK_version}   = $version;

    my $info = $info{$version};

    $self->compression(delete $args->{compression} ||COMPRESSION_LEVEL_DEFAULT);
    $self->format(delete $args->{format});

    push @{$args->{prefixes}}, @{$info->{prefixes} || []};

    unshift @{$args->{opts_readers}}
      , mixed_elements     => 'TEXTUAL'
      , sloppy_floats      => 1
      , sloppy_integers    => 1
      , hooks              => $info->{hooks_r};

    unshift @{$args->{opts_writers}}
      , hooks              => $info->{hooks_w};

    $self->SUPER::init($args);

    (my $xsd = __FILE__) =~ s,\.pm$,/xsd,;
    my @xsds = map +(glob "$xsd/$_"), @{$info->{schemas}};

    # don''t worry, XML::Compile::Schema will parse each file only once,
    # so only the first KML object created will consume considerable time.
    $self->importDefinitions(\@xsds);

    $self->declare(READER => 'kml', include_namespaces => 1);
    $self->declare(WRITER => 'kml');
    $self;
}


#-----------------------------


sub version() {shift->{GK_version}}
sub compression(;$)
{   my $self = shift;
    @_ ? ($self->{GK_compress} = shift) : $self->{GK_compress};
}
sub format(;$)
{   my $self = shift;
    @_ ? ($self->{GK_format} = shift) : $self->{GK_format};
}

#-----------------------------


sub writeKML($$;$)
{   my ($self, $data, $file, $zipped) = @_;

    my $doc    = XML::LibXML::Document->new('1.0', 'UTF-8');
    my $xml    = $self->writer('kml')->($doc, $data);
    $doc->setDocumentElement($xml);

    my $format = $self->format;
    $zipped ||= $file =~ m/\.kmz$/i;

    if($zipped)
    {   my $arch   = Archive::Zip->new;
        defined $format or $format = 0;
        my $member = $arch->addString($doc->toString($format), KML_NAME_IN_KMZ);
        $member->desiredCompressionLevel($self->compression);
        
        if(ref $file eq 'GLOB' || UNIVERSAL::isa($file, 'IO::Handle'))
        {   $arch->writeToFileHandle($file) == AZ_OK
                or fault __x"cannot write zip to filehandle";
        }
        else
        {   $arch->writeToFileNamed($file) == AZ_OK
                or fault __x"cannot write zip to {fn}", fn => $file;
        }
        return MIME_KMZ;
    }

    defined $format or $format = 1;
    if(ref $file eq 'GLOB' || UNIVERSAL::isa($file, 'IO::Handle'))
         { $doc->toFH  ($file, $format) }
    else { $doc->toFile($file, $format) }

    MIME_KML;
}

# name upto 0.02
sub readKML($;$)
{   my ($self, $data) = (shift, shift);
    @_ ? $self->from($data, is_compressed => shift) : $self->from($data);
}


sub from($@)
{   my ($class, $source, %args) = @_;
    my $zipped = exists $args{is_compressed}
               ? delete $args{is_compressed}
               : !ref $source && $source =~ m/\.kmz$/i;

    $source = $class->fromZipped($source)
        if $zipped;

    # if(open TR, '>', '/tmp/kml-trace')
    # {   print TR $root->toString(1);
    #     print TR "###\n\n";
    #     print TR ref $source ? $$source : $source;
    #     close TR;
    # }

    my ($root, %details) = XML::Compile->dataToXML($source);

    $root->nodeName eq 'kml'
        or error __x"content of {source} is not kml", source => $source;

    my $ns      = $root->namespaceURI;
    my $version = $ns2version{$ns}
        or error __x"kml type {ns} in {source} not supported (yet)"
             , ns => $ns, source => $source;

    my $kml     = $implement{$version} ||= $class->new(version => $version);
    ($ns, $kml->reader('kml', %args)->($root));
}

sub fromZipped($)
{   my ($class, $source) = @_;
    my $arch = Archive::Zip->new;

    # Archive::Zip can only read from files: it is very broken with
    # respect to IO::File objects and PerlIO.  So: when the data is
    # not a file, we have to create it.

    my ($fh, $filename, $type);
    if(!ref $source && $source !~ m/^\s*\</)
    {   # string not KML-XML hence filename
        ($type, $filename) = (file => $source);
    }
    else
    {   ($fh, $filename) = tempfile 'kmz-XXXXX'
          , UNLINK  => 1, TMPDIR  => 1, SUFFIX  => '.zip';

        if(!ref $source)                 # string with XML
        {   print $fh $source;
            $type = 'string';
        }
        elsif(ref $source eq 'SCALAR')   # ref-string with XML
        {   print $fh $$source;
            $type = 'scalar';
        }
        else      # let's hope it is a filehandle (compatible)
        {   print {$fh} <$source>;
            $type = 'filehandle';
        }
        seek $fh, 0, SEEK_SET;
    }

    $arch->read($filename)==AZ_OK
        or fault __x"cannot read zip headers from file {fn}, passed as {type}"
             , fn => $filename, type => $type;

    my $kml  = $arch->memberNamed(KML_NAME_IN_KMZ);
    my $buffer = '';
    open DOC, '>', \$buffer;
    $kml->extractToFileHandle(\*DOC) == AZ_OK
        or fault __x"failed extracting kml from zip {s}", s => $source;

    close DOC;
    \$buffer;
}

# IMO, the KML design makes a mistake in defining colors as hexBinary.
# The colors are integer values, in the program represented by things
# like 0xff34135, not binary blobs.  The following hooks make this work.
# Without those hooks, you would have to write pack("N", $color) all the
# time.

sub color_hex_read(@)
{   my ($elem, $reader, $path, $label, $replaced) = @_;

    my $text = $elem->textContent;
    $text =~ s/\s//g;

    my $value = unpack 'N', pack 'H8', $text;    # parse hex value into int.
    ($label => $value);
}

sub color_hex_write(@)
{   my ($doc, $value, $path, $label, $replaced) = @_;
    defined $value or return;  # for template

    my $node = $doc->createElement($label);
    my ($pv, $iv) = (DDual $value)[0,1];
    my $text = defined $iv  # integer value prevails
      ? (unpack 'H8', pack "N", $value)
      : $pv;   # validation by XML

    $node->appendText($text);
    $node;
}

1;
